File: Algorithms/src/nattybear/level1/GymSuit.js
function solution(n, lost, reserve) {

  const f = (x, y) => x - y;

  lost.sort(f);
  reserve.sort(f);

  function helper(predicate) {
    for (let i = 0; i < lost.length; i++) {
      for (let j = 0; j < reserve.length; j++) {
        let los = lost[i];
        let res = reserve[j];
        if (predicate(los, res)) {
          lost[i] = -1;
          reserve[j] = -1;
          break;
        }
      }
    }
  }

  helper((x, y) => x == y);
  helper((x, y) => Math.abs(x - y) == 1);

  const lost_ = lost.filter(x => x != -1);

  return n - lost_.length;
}

console.log(solution(5, [2,4], [1,3,5]));
console.log(solution(5, [2,4], [3]));
console.log(solution(3, [3], [1]));
console.log(solution(5, [2,3], [3,4]));

File: Algorithms/src/nattybear/level1/GymSuit.java
package nattybear.level1;

import java.util.Arrays;
import java.util.function.BiFunction;
import java.util.stream.IntStream;

public class GymSuit {
	private static class Solution {
		public int solution(int n, int[] losts, int[] reserves) {
			Arrays.sort(losts);
			Arrays.sort(reserves);

			Helper helper = new Helper(losts, reserves);

			helper.help((x, y) -> x == y);
			helper.help((x, y) -> Math.abs(x - y) == 1);

			IntStream stream = Arrays.stream(helper.getLosts());

			int numberOfLostStudent = (int) stream.filter(x -> x != -1).count();

			return n - numberOfLostStudent;
		}
	}

	private static class Helper {
		private int[] losts;
		private int[] reserves;

		public Helper(int[] losts, int[] reserves) {
			this.losts = losts;
			this.reserves = reserves;
		}

		public void help(BiFunction<Integer, Integer, Boolean> predicate) {
			for (int i = 0; i < losts.length; i++) {
				for (int j = 0; j < reserves.length; j++) {
					int lost = losts[i];
					int reserve = reserves[j];
					if (predicate.apply(lost, reserve)) {
						losts[i] = -1;
						reserves[j] = -1;
						break;
					}
				}
			}
		}

		public int[] getLosts() {
			return losts;
		}
	}

	public static void main(String[] args) {
		Solution s = new Solution();

		int[] losts = { 2, 4 };
		int[] reserves = { 1, 3, 5 };

		System.out.println(s.solution(5, losts, reserves));
	}
}
